src/server/model.coffee

```diff
@@ -486,12 +486,39 @@ module.exports = Model = (db, options) ->
     load docName, (error, doc) ->
       callback error, if doc then {v:doc.v, type:doc.type, snapshot:doc.snapshot, meta:doc.meta}
 
+  # Gets the snapshots for the specified document.
   @getSnapshots = (docname, callback) ->
     db.getSnapshots docname, (error, snapshots) ->
       if error
-        callback error, null
-        return
-      callback null, snapshots
+        callback? error
+      else
+        callback? null, snapshots
+
+  # Gets every version of the specified document, in intervals of `n`. This
+  # function doesn't use snapshot data, it rebuilds the document versions from
+  # the operations.
+  #
+  # Callback is called with (error, [{v: <version>, type: <type>, snapshot: <snapshot>, meta: <meta>}])
+  @getVersions = (docName, n, callback) ->
+    load docName, (error, doc) ->
+      return callback? error if error
+
+      # The the doc type of the latest snapshot.
+      type = doc.type
+
+      getOps docName, 0, null, (error, ops) ->
+        return callback? error if error
```

Comment:
Our production db (unfortunately) has a large number of documents (mostly from 2012-2015) where most operations have been deleted.

I haven't tested yet, buy I reckon this function will do weird things in that situation - maybe we should return an error and not even try?

Reply:
Yes, unfortunately the `getOps` function ignores missing ops in the database. So you can have a situation where rebuilding the document from the ops will fail.

I might see if I can load the ops directly from the db and check what the earliest version is. The internal API renumbers all the op versions starting from zero, even if zero doesn't actually exist in the DB 😮

Reply:
do the objects returned by `getOps` indicate a version number or is that just implied by the order of the objects?

Reply:
No, the line is actually [commented out](https://github.com/conversation/ShareJS/blob/master/src/server/db/pg.coffee#L171).

Reply:
🙄